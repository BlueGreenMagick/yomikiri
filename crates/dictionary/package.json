{
  "name": "@yomikiri/dictionary",
  "collaborators": [
    "BlueGreenMagick <bluegreenmagick@gmail.com>"
  ],
  "description": "Yomikiri dictionary files",
  "version": "0.1.0",
  "license": "MIT",
  "files": [
    "res/*"
  ],
  "scripts": {
    "wasm": "rm -rf ./pkg && wasm-pack build --scope yomikiri --target web",
    "test": "cargo test",
    "check": "cargo check --target \"$(rustc -vV | grep host | (IFS=\" \"; read -r key value; echo $value))\" --target wasm32-unknown-unknown",
    "lint": "cargo clippy --target \"$(rustc -vV | grep host | (IFS=\" \"; read -r key value; echo $value))\" --target wasm32-unknown-unknown -- -D warnings",
    "update-dictionary": "rm -rf ./jmdict && pnpm run rebuild",
    "rebuild": "cargo run --all-features --bin yomikiri-dictionary-builder",
    "download": "cargo run --all-features",
    "construct": "pnpm run download && pnpm run wasm"
  }
}