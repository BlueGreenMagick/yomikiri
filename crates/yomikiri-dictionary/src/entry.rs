//! Contains dictionary entry struct
//!
//! `Serialize`, `Deserialize`, `Tsify` is used when passing to web
//! `Decode`, `Encode` is used to encode the structs into the dictionary file

use std::ops::Deref;

use serde::de::Error as DeserializeError;
use serde::{Deserialize, Serialize};
use yomikiri_jmdict::jmdict::{JMDialect, JMPartOfSpeech, JMSenseMisc};
use yomikiri_unidic_types::{
    UnidicAdjectivePos2, UnidicInterjectionPos2, UnidicNaAdjectivePos2, UnidicNounPos2,
    UnidicParticlePos2, UnidicPos, UnidicSuffixPos2, UnidicSymbolPos2, UnidicVerbPos2,
};

#[cfg(feature = "wasm")]
use tsify_next::Tsify;

use crate::{Error, Result};

/// Constraints:
/// 1. Must have at least 1 reading
#[cfg_attr(feature = "wasm", derive(Tsify))]
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(transparent)]
pub struct Entry(EntryInner);

#[cfg_attr(feature = "wasm", derive(Tsify))]
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct EntryInner {
    pub id: u32,
    pub kanjis: Vec<Kanji>,
    pub readings: Vec<Reading>,
    pub grouped_senses: Vec<GroupedSense>,
    pub priority: u16,
}

#[cfg_attr(feature = "wasm", derive(Tsify))]
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Kanji {
    pub kanji: String,
    pub rarity: Rarity,
}

#[cfg_attr(feature = "wasm", derive(Tsify))]
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Reading {
    pub reading: String,
    /// This reading counts as term, not reading
    pub nokanji: bool,
    pub constrain: Vec<String>,
    pub rarity: Rarity,
}

/// Ordered by rarity.
/// `Normal` is the most common and `Search`` is the rarest.
#[cfg_attr(feature = "wasm", derive(Tsify))]
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
pub enum Rarity {
    Normal,
    Rare,
    Outdated,
    Incorrect,
    Search,
}

#[cfg_attr(feature = "wasm", derive(Tsify))]
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct GroupedSense {
    pub part_of_speech: Vec<PartOfSpeech>,
    pub senses: Vec<Sense>,
}

#[cfg_attr(feature = "wasm", derive(Tsify))]
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Sense {
    pub meanings: Vec<String>,
    // to_kanji or to_reading
    pub constrain: Vec<String>,
    pub misc: Vec<JMSenseMisc>,
    pub info: Vec<String>,
    pub dialects: Vec<JMDialect>,
}

pub type PartOfSpeech = JMPartOfSpeech;

fn jmpos_to_unidic(pos: &JMPartOfSpeech) -> UnidicPos {
    match pos {
        JMPartOfSpeech::Noun => UnidicPos::Noun(UnidicNounPos2::Unknown),
        JMPartOfSpeech::Verb => UnidicPos::Verb(UnidicVerbPos2::Unknown),
        JMPartOfSpeech::Adjective => UnidicPos::Adjective(UnidicAdjectivePos2::Unknown),
        JMPartOfSpeech::NaAdjective => UnidicPos::NaAdjective(UnidicNaAdjectivePos2::Unknown),
        JMPartOfSpeech::Particle => UnidicPos::Particle(UnidicParticlePos2::Unknown),
        JMPartOfSpeech::Adverb => UnidicPos::Adverb,
        JMPartOfSpeech::Interjection => UnidicPos::Interjection(UnidicInterjectionPos2::Unknown),
        JMPartOfSpeech::Suffix => UnidicPos::Suffix(UnidicSuffixPos2::Unknown),
        JMPartOfSpeech::AuxiliaryVerb => UnidicPos::AuxVerb,
        JMPartOfSpeech::Pronoun => UnidicPos::Pronoun,
        JMPartOfSpeech::Conjunction => UnidicPos::Conjunction,
        JMPartOfSpeech::Prefix => UnidicPos::Prefix,
        JMPartOfSpeech::Adnomial => UnidicPos::PrenounAdjectival,
        JMPartOfSpeech::Expression => UnidicPos::Expression,
        JMPartOfSpeech::Unclassified => UnidicPos::Unknown,
        JMPartOfSpeech::Symbol => UnidicPos::Symbol(UnidicSymbolPos2::Unknown),
    }
}

fn unidic_to_jmpos(unidic: &UnidicPos) -> JMPartOfSpeech {
    match unidic {
        UnidicPos::Noun(_) => JMPartOfSpeech::Noun,
        UnidicPos::Verb(_) => JMPartOfSpeech::Verb,
        UnidicPos::Adjective(_) => JMPartOfSpeech::Adjective,
        UnidicPos::NaAdjective(_) => JMPartOfSpeech::NaAdjective,
        UnidicPos::Particle(_) => JMPartOfSpeech::Particle,
        UnidicPos::Adverb => JMPartOfSpeech::Adverb,
        UnidicPos::Interjection(_) => JMPartOfSpeech::Interjection,
        UnidicPos::Suffix(_) => JMPartOfSpeech::Suffix,
        UnidicPos::AuxVerb => JMPartOfSpeech::AuxiliaryVerb,
        UnidicPos::Pronoun => JMPartOfSpeech::Pronoun,
        UnidicPos::Conjunction => JMPartOfSpeech::Conjunction,
        UnidicPos::Prefix => JMPartOfSpeech::Prefix,
        UnidicPos::PrenounAdjectival => JMPartOfSpeech::Adnomial,
        UnidicPos::Expression => JMPartOfSpeech::Expression,
        UnidicPos::SupplementarySymbol(_) => JMPartOfSpeech::Symbol,
        UnidicPos::Whitespace => JMPartOfSpeech::Symbol,
        UnidicPos::Symbol(_) => JMPartOfSpeech::Symbol,
        UnidicPos::Unknown => JMPartOfSpeech::Unclassified,
    }
}

impl Entry {
    pub fn new(inner: EntryInner) -> Result<Self> {
        Self::validate_entry(&inner)?;
        Ok(Self(inner))
    }

    fn validate_entry(inner: &EntryInner) -> Result<()> {
        if inner.readings.len() == 0 {
            return Err(Error::InvalidEntry(
                "Entry must contain at least 1 kanji or reading".into(),
            ));
        }
        Ok(())
    }

    pub fn main_form(&self) -> &str {
        if let Some(kanji) = self.kanjis.first() {
            if kanji.rarity == Rarity::Normal {
                return &kanji.kanji;
            }
        }
        if let Some(reading) = self.readings.first() {
            if reading.rarity == Rarity::Normal {
                return &reading.reading;
            }
        }
        if let Some(kanji) = self.kanjis.first() {
            return &kanji.kanji;
        }
        &self.readings.first().unwrap().reading
    }

    pub fn has_pos(&self, pos: PartOfSpeech) -> bool {
        self.grouped_senses
            .iter()
            .any(|g| g.part_of_speech.contains(&pos))
    }
}

impl Deref for Entry {
    type Target = EntryInner;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl<'de> Deserialize<'de> for Entry {
    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        let inner = EntryInner::deserialize(deserializer)?;
        Self::validate_entry(&inner).map_err(|e| D::Error::custom(e))?;

        Ok(Self(inner))
    }
}
