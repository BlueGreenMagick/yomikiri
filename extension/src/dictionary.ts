import pako from "pako";
import Dexie from "dexie";
import Utils from "~/utils";
import EnJMDict from "./assets/jmdict/en.json.gz";
import { Entry } from "./dicEntry";

/*
An entry can contain 0+ form, 1+ reading, and 1+ sense.

A reading can contain 0+ to_form. If 0, the reading is applied to all forms. 
If nokanji, it is associated with forms, but are not the true readings of the forms.
Else, only to specified forms.

A sense can contain 0+ to_reading, and 0+ to_form.

*/

// db schema version
const DB_VER = 1;
// dictionary schema version
const DIC_VER = 1;

interface Config {
  name: string;
  value: any;
}

export class Dictionary {
  db: Dexie;

  private constructor() {
    this.db = new Dexie("Dictionary");
    this.setupDB();
  }

  private static async downloadFromUrl(url: string): Promise<Entry[]> {
    const resp = await fetch(url);
    const data = await resp.arrayBuffer();
    const unzipped = pako.ungzip(data, { to: "string" }) as string;
    const entryObjects = JSON.parse(unzipped) as any[];
    return entryObjects.map(Entry.fromObject);
  }

  private setupDB() {
    // Should we use multi-valued keys instead of sql-like many-to-many?
    this.db.version(DB_VER).stores({
      entries: "++id, *terms",
      config: "&name",
    });
  }

  private async install(url: string): Promise<void> {
    Utils.bench("Installing dictionary");
    const db = this.db;
    const entries = await Dictionary.downloadFromUrl(url);
    await db.transaction("rw", "entries", "config", async () => {
      await db.table("config").put({ name: "dic_ver", value: DIC_VER });
      await db.table("config").put({ name: "dic_name", value: "English" });
      await db.table("entries").clear();
      await db.table("entries").bulkAdd(entries);
    });
    Utils.bench("Installed dictionary");
  }

  private async getConfig(name: string): Promise<any> {
    return await this.db.table("config").get(name);
  }

  static async initialize(): Promise<Dictionary> {
    const dic = new Dictionary();
    const dic_ver = await dic.getConfig("dic_ver");
    if (!(typeof dic_ver === "object" && dic_ver.value === DIC_VER)) {
      await dic.install(EnJMDict);
    }
    return dic;
  }

  async search(term: string): Promise<Entry[]> {
    return await this.db.table("entries").where("terms").equals(term).toArray();
  }

  async hasStartsWith(
    term: string,
    filter?: (entry: Entry) => boolean
  ): Promise<boolean> {
    if (!filter) {
      filter = (e: Entry) => true;
    }
    const entry = await this.db
      .table("entries")
      .where("terms")
      .startsWith(term)
      .distinct()
      .filter(filter)
      .first();
    return entry !== undefined;
  }
}
