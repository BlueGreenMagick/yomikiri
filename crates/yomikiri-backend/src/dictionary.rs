use anyhow::{Context, Result};
use yomikiri_dictionary::dictionary::{Dictionary as InnerDictionary, DictionaryMetadata};
use yomikiri_dictionary::entry::{Entry, NameEntry, Rarity, WordEntry};
use yomikiri_dictionary::PartOfSpeech;

use crate::tokenize::InnerToken;

pub struct Dictionary<D: AsRef<[u8]> + 'static> {
    inner: InnerDictionary<D>,
}

impl<D: AsRef<[u8]> + 'static> Dictionary<D> {
    pub fn try_new(source: D) -> Result<Dictionary<D>> {
        let inner =
            InnerDictionary::try_decode(source).context("Could not read dictionary file.")?;
        Ok(Self { inner })
    }

    pub fn search(&self, term: &str) -> Result<Vec<Entry>> {
        let view = self.inner.borrow_view();
        let terms = &view.term_index;
        let idxs = terms.get(term)?;
        let entries = view.get_entries(&idxs)?;
        Ok(entries)
    }

    pub fn search_meaning(&self, query: &str) -> Result<Vec<Entry>> {
        let view = self.inner.borrow_view();
        let entries = view.search_meaning(query)?;
        Ok(entries)
    }

    /// Returns true only if there is a dictionary term
    /// that starts with `prefix` and is not `prefix`
    pub fn has_starts_with_excluding(&self, prefix: &str) -> bool {
        // assumes there is at least 1 entry.
        // needed in order to create a bytestring that is 1 greater than prefix below
        if prefix.is_empty() {
            return true;
        }

        let mut next_prefix_bytes = prefix.as_bytes().to_vec();
        let len = next_prefix_bytes.len();
        next_prefix_bytes[len - 1] += 1;

        let terms = &self.inner.borrow_view().term_index;
        terms.has_starts_with_excluding(prefix)
    }

    pub fn contains(&self, term: &str) -> bool {
        self.inner.borrow_view().term_index.contains_key(term)
    }

    /// Returns json text of entries
    pub fn search_json(&self, term: &str) -> Result<Vec<String>> {
        let view = &self.inner.borrow_view();
        let terms = &self.inner.borrow_view().term_index;
        let idxs = terms.get(term)?;
        let entries = view.get_entries(&idxs)?;
        let jsons = entries
            .iter()
            .map(serde_json::to_string)
            .collect::<serde_json::Result<Vec<String>>>()
            .context("Failed to serialize dictionary entry into JSON")?;
        Ok(jsons)
    }

    pub fn metadata(&self) -> &DictionaryMetadata {
        &self.inner.borrow_view().metadata
    }

    /// Finds entries, ordered by what best matches token
    ///
    /// 0. Name entry -> Word entry
    /// 1. Non-search -> search-only
    /// 2. token.base -> token.text
    /// 3. Entries whose POS matches token.pos
    /// 4. Rare -> Non rare
    /// 5. Entry with higher priority is shown first

    pub(crate) fn search_for_token(&self, token: &InnerToken) -> Result<Vec<Entry>> {
        struct EntryMeta {
            entry: WordEntry,
            rarity: Rarity,
            from_base: bool,
        }

        let mut name_entries: Vec<NameEntry> = vec![];
        // word entry metas
        let mut entry_metas: Vec<EntryMeta> = vec![];

        let entries = self.search(&token.base)?;
        for entry in entries {
            match entry {
                Entry::Word(entry) => {
                    let rarity = entry.term_rarity(&token.base)?;
                    let entry_meta = EntryMeta {
                        entry,
                        rarity,
                        from_base: true,
                    };
                    entry_metas.push(entry_meta);
                }
                Entry::Name(entry) => {
                    name_entries.push(entry);
                }
            }
        }

        let entries = self.search(&token.text)?;
        for entry in entries {
            match entry {
                Entry::Word(entry) => {
                    if entry_metas.iter().any(|e| e.entry == entry) {
                        continue;
                    }
                    let rarity = entry.term_rarity(&token.text)?;
                    let entry_meta = EntryMeta {
                        entry,
                        rarity,
                        from_base: false,
                    };
                    entry_metas.push(entry_meta);
                }
                Entry::Name(entry) => {
                    if name_entries.contains(&entry) {
                        continue;
                    }
                    name_entries.push(entry);
                }
            }
        }

        // Sort entries. Less means 'a' comes before 'b'
        entry_metas.sort_by(|a, b| {
            let a_is_search = a.rarity == Rarity::Search;
            let b_is_search = b.rarity == Rarity::Search;
            a_is_search
                .cmp(&b_is_search)
                .then(a.from_base.cmp(&b.from_base).reverse())
                .then_with(|| {
                    let pos = PartOfSpeech::from(&token.pos);
                    a.entry.has_pos(pos).cmp(&b.entry.has_pos(pos)).reverse()
                })
                .then_with(|| {
                    let a_is_normal = a.rarity == Rarity::Normal;
                    let b_is_normal = a.rarity == Rarity::Normal;
                    a_is_normal.cmp(&b_is_normal).reverse()
                })
                .then(a.entry.priority.cmp(&b.entry.priority).reverse())
        });

        let mut entries: Vec<Entry> = name_entries.into_iter().map(Entry::Name).collect();
        for meta in entry_metas {
            entries.push(Entry::Word(meta.entry));
        }
        Ok(entries)
    }
}
