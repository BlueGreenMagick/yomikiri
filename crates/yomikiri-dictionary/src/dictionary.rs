use std::io::{BufRead, Write};

use bincode::Options;
use ouroboros::self_referencing;
use serde::{Deserialize, Serialize};
use yomikiri_jmdict::{JMDictParser, JMneDictParser};

use crate::entry::{Entry, NameEntry};
use crate::index::{create_sorted_term_indexes, DictIndexMap, EntryIdx};
use crate::jagged_array::JaggedArray;
use crate::jmnedict::{parse_jmnedict_entry, NameEntriesBuilder};
use crate::{Result, WordEntry};

#[cfg(feature = "wasm")]
use tsify_next::Tsify;

#[self_referencing]
pub struct Dictionary<D: AsRef<[u8]> + 'static> {
    source: Box<D>,
    #[borrows(source)]
    #[covariant]
    pub view: DictionaryView<'this>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "wasm", derive(Tsify))]
#[cfg_attr(feature = "wasm", tsify(into_wasm_abi))]
pub struct DictionaryMetadata {
    jmdict_creation_date: Option<String>,
    jmnedict_creation_date: Option<String>,
}

pub struct DictionaryView<'a> {
    pub term_index: DictIndexMap<'a, EntryIdx>,
    pub entries: JaggedArray<'a, WordEntry>,
    pub name_entries: JaggedArray<'a, NameEntry>,
    pub metadata: DictionaryMetadata,
}

impl<D: AsRef<[u8]> + 'static> Dictionary<D> {
    pub fn try_decode(source: D) -> Result<Self> {
        let builder = DictionaryTryBuilder {
            source: Box::new(source),
            view_builder: |source| {
                DictionaryView::try_decode(source.as_ref().as_ref()).map(|(inner, _len)| inner)
            },
        };
        builder.try_build()
    }
}

impl<'a> DictionaryView<'a> {
    pub fn try_decode(source: &'a [u8]) -> Result<(Self, usize)> {
        let mut at = 0;
        let (term_index, len) = DictIndexMap::try_decode(&source[at..])?;
        at += len;
        let (entries, len) = JaggedArray::try_decode(&source[at..])?;
        at += len;
        let (name_entries, len) = JaggedArray::try_decode(&source[at..])?;
        at += len;
        let metadata = bincode::options().deserialize_from(&source[at..])?;
        let s = Self {
            name_entries,
            term_index,
            entries,
            metadata,
        };
        Ok((s, at))
    }

    pub fn build_and_encode_to<W: Write>(
        name_entries: &[NameEntry],
        entries: &[WordEntry],
        metadata: &DictionaryMetadata,
        writer: &mut W,
    ) -> Result<()> {
        let term_index_items = create_sorted_term_indexes(name_entries, entries)?;
        DictIndexMap::build_and_encode_to(&term_index_items, writer)?;
        JaggedArray::build_and_encode_to(entries, writer)?;
        JaggedArray::build_and_encode_to(name_entries, writer)?;
        bincode::options().serialize_into(writer, metadata)?;
        Ok(())
    }

    pub fn get_entries(&self, pointers: &[EntryIdx]) -> Result<Vec<Entry>> {
        pointers
            .iter()
            .map(|p| match p {
                EntryIdx::Word(idx) => self.entries.get(idx.0 as usize).map(Entry::Word),
                EntryIdx::Name(idx) => self.name_entries.get(idx.0 as usize).map(Entry::Name),
            })
            .collect::<Result<Vec<Entry>>>()
    }
}

pub struct DictionaryWriterJMDict {}
pub struct DictionaryWriterJMneDict {
    entries: Vec<WordEntry>,
    jmdict_creation_date: Option<String>,
}

pub struct DictionaryWriterFinal {
    entries: Vec<WordEntry>,
    jmdict_creation_date: Option<String>,
    name_entries: Vec<NameEntry>,
    jmnedict_creation_date: Option<String>,
}

#[derive(Default)]
pub struct DictionaryWriter<STATE> {
    state: STATE,
}

impl DictionaryWriter<DictionaryWriterJMDict> {
    pub fn new() -> Self {
        Self {
            state: DictionaryWriterJMDict {},
        }
    }

    pub fn read_jmdict<R: BufRead>(
        self,
        jmdict: R,
    ) -> Result<DictionaryWriter<DictionaryWriterJMneDict>> {
        let mut parser = JMDictParser::new(jmdict)?;
        // 203736 entries (2022-08-23)
        let mut entries = Vec::with_capacity(210000);
        while let Some(entry) = parser.next_entry()? {
            if entry.id < 5000000 || entry.id >= 6000000 {
                entries.push(WordEntry::try_from(entry)?)
            }
        }
        Ok(DictionaryWriter {
            state: DictionaryWriterJMneDict {
                entries,
                jmdict_creation_date: parser.creation_date().map(|d| d.to_string()),
            },
        })
    }
}

impl DictionaryWriter<DictionaryWriterJMneDict> {
    pub fn read_jmnedict<R: BufRead>(
        mut self,
        jmnedict: R,
    ) -> Result<DictionaryWriter<DictionaryWriterFinal>> {
        let mut name_builder = NameEntriesBuilder::new();

        let mut parser = JMneDictParser::new(jmnedict)?;
        while let Some(entry) = parser.next_entry()? {
            parse_jmnedict_entry(&mut self.state.entries, &mut name_builder, entry)?;
        }

        let mut name_entries = Vec::with_capacity(NameEntriesBuilder::ENTRIES_CAPACITY);
        for name_entry in name_builder.into_iter() {
            name_entries.push(name_entry);
        }

        Ok(DictionaryWriter {
            state: DictionaryWriterFinal {
                entries: self.state.entries,
                jmdict_creation_date: self.state.jmdict_creation_date,
                name_entries,
                jmnedict_creation_date: parser.creation_date().map(|d| d.to_string()),
            },
        })
    }
}

impl DictionaryWriter<DictionaryWriterFinal> {
    pub fn write<W: Write>(self, writer: &mut W) -> Result<()> {
        let metadata = DictionaryMetadata {
            jmdict_creation_date: self.state.jmdict_creation_date,
            jmnedict_creation_date: self.state.jmnedict_creation_date,
        };
        DictionaryView::build_and_encode_to(
            &self.state.name_entries,
            &self.state.entries,
            &metadata,
            writer,
        )?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use yomikiri_jmdict::jmnedict::JMneNameType;

    use crate::dictionary::{Dictionary, DictionaryMetadata, DictionaryView};
    use crate::entry::{
        GroupedNameItem, Kanji, NameEntry, NameItem, Rarity, Reading, WordEntry, WordEntryInner,
    };
    use crate::Result;

    #[test]
    fn write_then_read_dictionary_with_single_word_entry() -> Result<()> {
        let inner = WordEntryInner {
            id: 1234,
            kanjis: vec![
                Kanji {
                    kanji: "読み切り".into(),
                    rarity: Rarity::Normal,
                },
                Kanji {
                    kanji: "読みきり".into(),
                    rarity: Rarity::Normal,
                },
            ],
            readings: vec![Reading {
                reading: "よみきり".into(),
                nokanji: false,
                to_kanji: vec![],
                rarity: Rarity::Normal,
            }],
            grouped_senses: vec![],
            priority: 10,
        };
        let entry = WordEntry::new(inner)?;
        let mut buffer = Vec::with_capacity(1024);
        let metadata = DictionaryMetadata {
            jmdict_creation_date: Some("2024-09-19".into()),
            jmnedict_creation_date: None,
        };
        DictionaryView::build_and_encode_to(&[], &[entry.clone()], &metadata, &mut buffer)?;
        let dict = Dictionary::try_decode(buffer)?;

        let view = dict.borrow_view();
        assert_eq!(view.name_entries.len(), 0);
        assert_eq!(view.entries.len(), 1);
        assert_eq!(view.entries.get(0)?, entry);
        Ok(())
    }

    #[test]
    fn write_then_read_dictionary_with_single_name_entry() -> Result<()> {
        let entry = NameEntry {
            kanji: "雅哉".into(),
            groups: vec![GroupedNameItem {
                types: vec![JMneNameType::Forename],
                items: vec![NameItem {
                    id: 5174270,
                    reading: "まさや".into(),
                }],
            }],
        };
        let metadata = DictionaryMetadata {
            jmdict_creation_date: None,
            jmnedict_creation_date: Some("2024-09-19".into()),
        };
        let mut buffer = Vec::with_capacity(1024);
        DictionaryView::build_and_encode_to(&[entry.clone()], &[], &metadata, &mut buffer)?;
        let dict = Dictionary::try_decode(buffer)?;

        let view = dict.borrow_view();
        assert_eq!(view.entries.len(), 0);
        assert_eq!(view.name_entries.len(), 1);
        assert_eq!(view.name_entries.get(0)?, entry);
        Ok(())
    }
}
