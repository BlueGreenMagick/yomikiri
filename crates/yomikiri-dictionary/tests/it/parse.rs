use itertools::Itertools;
use yomikiri_dictionary::dictionary::{DictionaryView, DictionaryWriter};
use yomikiri_dictionary::entry::NameEntry;
use yomikiri_dictionary::WordEntry;

type Result<T> = std::result::Result<T, Box<dyn std::error::Error>>;

fn parse_jmdicts(
    jmdict: Option<&str>,
    jmnedict: Option<&str>,
) -> Result<(Vec<WordEntry>, Vec<NameEntry>)> {
    let writer = DictionaryWriter::new();
    let jmdict = jmdict.unwrap_or("<JMdict></JMdict>");
    let jmnedict = jmnedict.unwrap_or("<JMnedict></JMnedict>");

    let writer = writer.read_jmdict(jmdict.as_bytes())?;
    let writer = writer.read_jmnedict(jmnedict.as_bytes())?;
    let mut bytes = Vec::with_capacity(128);
    writer.write(&mut bytes)?;

    let (dict, _) = DictionaryView::try_decode(&bytes)?;
    let mut word_entries: Vec<WordEntry> = dict.entries.all_items_iter().try_collect()?;
    let mut name_entries: Vec<NameEntry> = dict.name_entries.all_items_iter().try_collect()?;
    word_entries.sort_by(|a, b| a.id.cmp(&b.id));
    name_entries.sort_by(|a, b| a.kanji.cmp(&b.kanji));

    Ok((word_entries, name_entries))
}

#[test]
fn test_jmdict_parse_1() -> Result<()> {
    let xml = r#"<?xml version="1.0" encoding="UTF-8"?>
<!-- JMdict created: 2024-08-23 -->
<JMdict>
<entry>
<ent_seq>1000810</ent_seq>
<k_ele>
<keb>いじり回す</keb>
</k_ele>
<k_ele>
<keb>弄り回す</keb>
</k_ele>
<k_ele>
<keb>弄りまわす</keb>
</k_ele>
<r_ele>
<reb>いじりまわす</reb>
</r_ele>
<sense>
<pos>&v5s;</pos>
<pos>&vt;</pos>
<gloss>to tinker with</gloss>
<gloss>to fumble with</gloss>
<gloss>to fiddle with</gloss>
<gloss>to twiddle</gloss>
</sense>
</entry>
</JMdict>
"#;
    let result = parse_jmdicts(Some(xml), None)?;
    insta::assert_yaml_snapshot!(result);
    Ok(())
}

#[test]
fn test_jmdict_parse_2() -> Result<()> {
    let xml = r#"<?xml version="1.0" encoding="UTF-8"?>
<!-- JMdict created: 2024-08-23 -->
<JMdict>
<entry>
<ent_seq>1006950</ent_seq>
<k_ele>
<keb>抑</keb>
<ke_inf>&rK;</ke_inf>
</k_ele>
<k_ele>
<keb>抑々</keb>
<ke_inf>&rK;</ke_inf>
</k_ele>
<k_ele>
<keb>抑抑</keb>
<ke_inf>&rK;</ke_inf>
</k_ele>
<k_ele>
<keb>抑も</keb>
<ke_inf>&io;</ke_inf>
<ke_inf>&rK;</ke_inf>
</k_ele>
<r_ele>
<reb>そもそも</reb>
</r_ele>
<sense>
<pos>&n;</pos>
<pos>&adv;</pos>
<misc>&uk;</misc>
<gloss>in the first place</gloss>
</sense>
<sense>
<pos>&conj;</pos>
<misc>&uk;</misc>
<s_inf>used when bringing up something already mentioned</s_inf>
<gloss>after all</gloss>
<gloss>anyway</gloss>
</sense>
<sense>
<pos>&n;</pos>
<pos>&adj-no;</pos>
<misc>&uk;</misc>
<gloss>beginning</gloss>
<gloss>start</gloss>
</sense>
</entry>
</JMdict>
"#;

    let result = parse_jmdicts(Some(xml), None)?;
    insta::assert_yaml_snapshot!(result);
    Ok(())
}

#[test]
fn skip_name_entries_in_jmdict() -> Result<()> {
    let xml = r#"<?xml version="1.0"?>
<!-- JMnedict created: 2024-08-29 -->
<JMdict>
<entry>
<ent_seq>2862076</ent_seq>
<k_ele>
<keb>自己査定</keb>
</k_ele>
<r_ele>
<reb>じこさてい</reb>
</r_ele>
<sense>
<pos>&n;</pos>
<gloss>self-assessment</gloss>
<gloss>self-evaluation</gloss>
<gloss>self-appraisal</gloss>
<gloss>self-audit</gloss>
</sense>
</entry>
<entry>
<ent_seq>5000058</ent_seq>
<k_ele>
<keb>ＰＣエンジン</keb>
</k_ele>
<r_ele>
<reb>ピーシーエンジン</reb>
</r_ele>
<sense>
<pos>&n;</pos>
<misc>&product;</misc>
<gloss>(NEC) PC Engine</gloss>
<gloss>TurboGrafx-16</gloss>
</sense>
</entry>
</JMdict>
"#;
    let result = parse_jmdicts(Some(xml), None)?;
    insta::assert_yaml_snapshot!(result);
    Ok(())
}

#[test]
fn test_jmnedict_parse_single() -> Result<()> {
    let xml = r#"<?xml version="1.0"?>
<!-- JMnedict created: 2024-08-29 -->
<JMnedict>
<entry>
<ent_seq>5073921</ent_seq>
<k_ele>
<keb>恵山</keb>
</k_ele>
<r_ele>
<reb>ヘサン</reb>
</r_ele>
<r_ele>
<reb>けいざん</reb>
</r_ele>
<trans>
<name_type>&place;</name_type>
<trans_det>Hyesan (North Korea)</trans_det>
</trans>
</entry>
</JMnedict>
"#;
    let result = parse_jmdicts(None, Some(xml))?;
    insta::assert_yaml_snapshot!(result);
    Ok(())
}

// multiple jmne entries are merged into one name entry
// and also creates a word entry
#[test]
fn test_jmnedict_parse_multiple() -> Result<()> {
    let xml = r#"<?xml version="1.0"?>
<!-- JMnedict created: 2024-08-29 -->
<JMnedict>
<entry>
<ent_seq>5000268</ent_seq>
<k_ele>
<keb>あく屋</keb>
</k_ele>
<r_ele>
<reb>あくや</reb>
</r_ele>
<trans>
<name_type>&place;</name_type>
<trans_det>Akuya</trans_det>
</trans>
</entry>
<entry>
<ent_seq>5538229</ent_seq>
<k_ele>
<keb>鏑木</keb>
</k_ele>
<r_ele>
<reb>かぶらき</reb>
</r_ele>
<trans>
<name_type>&place;</name_type>
<name_type>&surname;</name_type>
<trans_det>Kaburaki</trans_det>
</trans>
</entry>
<entry>
<ent_seq>5538230</ent_seq>
<k_ele>
<keb>鏑木</keb>
</k_ele>
<r_ele>
<reb>かぶらぎ</reb>
</r_ele>
<trans>
<name_type>&surname;</name_type>
<trans_det>Kaburagi</trans_det>
</trans>
</entry>
<entry>
<ent_seq>5057716</ent_seq>
<r_ele>
<reb>ナウシカア</reb>
<re_pri>spec1</re_pri>
</r_ele>
<r_ele>
<reb>ナウシカー</reb>
</r_ele>
<trans>
<name_type>&fem;</name_type>
<trans_det>Nausicaa</trans_det>
</trans>
<trans>
<name_type>&char;</name_type>
<trans_det>Nausicaa (in Homer's Odyssey)</trans_det>
</trans>
</entry>
<entry>
<ent_seq>5538231</ent_seq>
<k_ele>
<keb>鏑木</keb>
</k_ele>
<r_ele>
<reb>かぶらやき</reb>
</r_ele>
<trans>
<name_type>&surname;</name_type>
<trans_det>Kaburayaki</trans_det>
</trans>
</entry>
</JMnedict>
"#;

    let result = parse_jmdicts(None, Some(xml))?;
    insta::assert_yaml_snapshot!(result);

    Ok(())
}
