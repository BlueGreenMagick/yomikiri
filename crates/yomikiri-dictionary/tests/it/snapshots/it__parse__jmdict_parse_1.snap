---
source: crates/yomikiri-dictionary/tests/it/parse.rs
assertion_line: 62
expression: result
---
- - id: 1000810
    kanjis:
      - kanji: いじり回す
        rarity: normal
      - kanji: 弄り回す
        rarity: normal
      - kanji: 弄りまわす
        rarity: normal
    readings:
      - reading: いじりまわす
        nokanji: false
        toKanji: []
        rarity: normal
    groupedSenses:
      - pos:
          - verb
        senses:
          - meanings:
              - to tinker with
              - to fumble with
              - to fiddle with
              - to twiddle
            toKanji: []
            toReading: []
            misc: []
            info: []
            dialects: []
    priority: 0
- []
